src/java/org/apache/cassandra/schema/PartitionDenylist.java

```diff
@@ -0,0 +1,462 @@
+                if (!endToken.isMinimum())
+                    subSet = subSet.headSet(endToken, PartitionPosition.Kind.MAX_BOUND.equals(range.right.kind()));
+                return subSet.size();
+            }
+
+            // Wrap around case
+            return denylistEntry.tokens.tailSet(startToken, PartitionPosition.Kind.MIN_BOUND.equals(range.left.kind())).size()
+                   + denylistEntry.tokens.headSet(endToken, PartitionPosition.Kind.MAX_BOUND.equals(range.right.kind())).size();
+        }
+        catch (final ExecutionException e)
+        {
+            logAccessFailure(tmd, e);
+            return 0;
+        }
+    }
+
+    private DenylistEntry getDenylistForTable(final TableId tid)
+    {
+        final TableMetadata tmd = Schema.instance.getTableMetadata(tid);
+        if (tmd == null)
+            return new DenylistEntry();
+
+        // We pull max keys + 1 in order to check below whether we've surpassed the allowable limit or not
+        final String readDenylist = String.format("SELECT * FROM %s.%s WHERE ks_name='%s' AND cf_name='%s' LIMIT %d", SystemDistributedKeyspace.NAME, PARTITION_DENYLIST_TABLE,
+                                                  tmd.keyspace, tmd.name, DatabaseDescriptor.getMaxDenylistKeysPerTable() + 1);
+
+        try
+        {
+            final UntypedResultSet results = process(readDenylist, DatabaseDescriptor.getDenylistConsistencyLevel());
+            if (results == null || results.isEmpty())
+                return new DenylistEntry();
+
+            if (results.size() > DatabaseDescriptor.getMaxDenylistKeysPerTable())
+                logger.error("Partition denylist for {}/{} has exceeded the maximum allowable size ({}). Remaining keys were ignored; please reduce the number of keys denied or increase the size of your cache to avoid inconsistency in denied partitions across nodes.", tmd.keyspace, tmd.name, DatabaseDescriptor.getMaxDenylistKeysPerTable());
+
+            final Set<ByteBuffer> keys = new HashSet<>();
+            final NavigableSet<Token> tokens = new TreeSet<>();
+            for (final UntypedResultSet.Row row : results)
+            {
+                final ByteBuffer key = row.getBlob("key");
+                keys.add(key);
+                tokens.add(StorageService.instance.getTokenMetadata().partitioner.getToken(key));
+            }
+            return new DenylistEntry(ImmutableSet.copyOf(keys), ImmutableSortedSet.copyOf(tokens));
+        }
+        catch (final RequestExecutionException e)
+        {
+            logger.error("Error reading partition_denylist table for {}/{}. Returning empty list.", tmd.keyspace, tmd.name, e);
+            return new DenylistEntry();
+        }
+    }
+
+    /**
+     * We check in this method whether either a table or the sum of all denylist keys surpass our allowable limits and
+     * warn to the user if we're above our threshold.
+     */
+    private Map<TableId, DenylistEntry> getDenylistForAllTables()
```

Comment:
This methods seems to not honor `getMaxDenylistKeysPerTable`. we load data like this
t1: 999 records
t2: 1 record 

for 500 per table, 1000 total limit.

Reply:
In its current form it does appear to allow overflow on a per table basis while respecting the max value, yes - good catch. I'll work on this tomorrow morning; have another implementation in mind that should resolve this.